src/main/kotlin/kweb/state/render.kt

```diff
@@ -33,15 +36,48 @@ fun <T : Any?> ElementCreator<*>.render(
 
     fun eraseAndRender() {
         do {
-            containerElement.removeChildren()
-            containerElement.new {
-                previousElementCreator.getAndSet(this)?.cleanup()
-                renderState.set(RENDERING_NO_PENDING_CHANGE)
-                block(value.value)
-                if (renderState.get() == RENDERING_NO_PENDING_CHANGE) {
-                    renderState.set(NOT_RENDERING)
+            val outboundMessages: List<JsFunction>? = if (parent.browser.kweb.isCatchingOutbound() == false) {
+                parent.browser.kweb.catchOutbound {
+                    containerElement.removeChildren()
+                    containerElement.new {
+                        previousElementCreator.getAndSet(this)?.cleanup()
+                        renderState.set(RENDERING_NO_PENDING_CHANGE)
+                        block(value.value)
+                        if (renderState.get() == RENDERING_NO_PENDING_CHANGE) {
+                            renderState.set(NOT_RENDERING)
+                        }
+                    }
                 }
+            } else {
+                containerElement.removeChildren()
+                containerElement.new {
+                    previousElementCreator.getAndSet(this)?.cleanup()
+                    renderState.set(RENDERING_NO_PENDING_CHANGE)
+                    block(value.value)
+                    if (renderState.get() == RENDERING_NO_PENDING_CHANGE) {
+                        renderState.set(NOT_RENDERING)
+                    }
+                }
+                null
+            }
+            if (outboundMessages != null) {
+                /*TODO to make this work I had to make WebBrowser.sessionId and Kweb.clientState public instead of private
+                  I think we may want to write a send() method in Kweb that we can call here, and set those vals back to private
```

Comment:
Ah, I guess this answers my previous question.

Reply:
Now that I think about it, I could also make public getters for these two fields and set the val's back to private. Should we go this route?

Reply:
If they need to be publicly accessible then I don't think there would be any advantage to using separate getters, a `val` is essentially a getter.